### DISCUSSION TITLE
Help with custom loss function

### DISCUSSION THREAD

**Discussion by ccclalala123** [MilesCranmer/PySR#676] [Q&A] [answered]
Hi,

I am trying to create a custom loss function based on the dataset. But I am not sure how to use this way in the julia loss function or whether it would be even possible.

Any help would be greatly appreciated!

Current Code:

```
import numpy as np

t1 = []
t2 = []
dudx = []
dvdy = []
S11 = []
S22 = []
S12_Real = []
S13_Real = []

filename = r'C:\Users\lenovo\Desktop\data.dat'
with open(filename, 'r') as file:
    for line in file:
        columns=line.split()
        t1.append(float(columns[0]))
        t2.append(float(columns[1]))
        dudx.append(float(columns[2]))
        dvdy.append(float(columns[3]))
        S11.append(float(columns[4]))
        S22.append(float(columns[5]))
        S12_Real.append(float(columns[6]))
        S13_Real.append(float(columns[7]))

t1 = np.array(t1)
t2 = np.array(t2)
dudx = np.array(dudx)
dvdy = np.array(dvdy)
S11 = np.array(S11)
S22 = np.array(S22)
S12_Real = np.array(S11)
S13_Real = np.array(S22)

X=np.zeros((len(t1),2))
X[:,0]=t1
X[:,1]=t2

from pysr import PySRRegressor

objective = """
function eval_loss(tree, dataset::Dataset, options)

    prediction, flag = eval_tree_array(tree, dataset.X, options)
    if !flag
        return Inf
    end

    S12 = (1.0 .- prediction) .* (S11 ./ dudx)
    S13 = (1.0 .- prediction) .* (S22 ./ dvdy)

    diffs1 = S12 - S12_Real
    diffs2 = S13 - S13_Real

    error1 = sum(diffs1 .^ 2) / length(diffs1)
    error2 = sum(diffs2 .^ 2) / length(diffs2)
    
    error = error1 + error2

    return error
end
"""

model = PySRRegressor(
    niterations=1000,
    batch_size=1000,
    binary_operators=[ "+", "-", "*", "/"],
    loss_function=objective,
)
model.fit(X, y)
model.get_best().equation
```

**Comment by MilesCranmer** [comment, marked as answer]
Thanks for the post. I think what you will want to do is create a constant in Julia. You can see my answer here: https://github.com/MilesCranmer/PySR/discussions/668#discussioncomment-10039206 for a related tip. I think this will apply here as well – you would create a constant for `S11` and `S22`.

**Reply by ccclalala123** [reply]
It works! Thank you so much!
